src/core_atmosphere/physics/physics_wrf/module_bl_ysu.F

```diff
@@ -110,1655 +132,343 @@ subroutine ysu(u3d,v3d,th3d,t3d,qv3d,qc3d,qi3d,p3d,p3di,pi3d,               &
 
-      INTEGER,INTENT(IN) :: KTS,KTE
-      REAL, INTENT(OUT) :: zi
-      REAL, INTENT(IN) :: landsea
-      REAL, DIMENSION(KTS:KTE), INTENT(IN) :: thetav1D, qke1D, dz1D
-      REAL, DIMENSION(KTS:KTE+1), INTENT(IN) :: zw1D
-      !LOCAL VARS
-      REAL ::  PBLH_TKE,qtke,qtkem1,wt,maxqke,TKEeps,minthv
-      REAL :: delt_thv   !delta theta-v; dependent on land/sea point
-      REAL, PARAMETER :: sbl_lim  = 200. !Theta-v PBL lower limit of trust (m).
-      REAL, PARAMETER :: sbl_damp = 400. !Damping range for averaging with TKE-based PBLH (m).
-      INTEGER :: I,J,K,kthv,ktke
+      do n = 1, nmix
+         do k = kts, kte
+            do i = its, ite
+               qmix_hv(i,k,n) = qmix(i,k,j,n)
+            end do
+         end do
+      end do
 
-      !FIND MAX TKE AND MIN THETAV IN THE LOWEST 500 M
-      k = kts+1
-      kthv = 1
-      ktke = 1
-      maxqke = 0.
-      minthv = 9.E9
+      do k = kts, kte+1
+         do i = its, ite
+            p3di_hv(i,k) = p3di(i,k,j)
+         end do
+      end do
 
-      DO WHILE (zw1D(k) .LE. 500.)
-        qtke  =MAX(Qke1D(k),0.)   ! maximum QKE
-         IF (maxqke < qtke) then
-            maxqke = qtke
-            ktke = k
-         ENDIF
-         IF (minthv > thetav1D(k)) then
-             minthv = thetav1D(k)
-             kthv = k
-         ENDIF
-         k = k+1
-      ENDDO
-      !TKEeps = maxtke/20. = maxqke/40.
-      TKEeps = maxqke/40.
-      TKEeps = MAX(TKEeps,0.025)
-      TKEeps = MIN(TKEeps,0.25)
+      do k = kts, kte
+         do i = its, ite
+            u3d_hv(i,k) = u3d(i,k,j)
+            v3d_hv(i,k) = v3d(i,k,j)
+            t3d_hv(i,k) = t3d(i,k,j)
+            qv3d_hv(i,k) = qv3d(i,k,j)
+            qc3d_hv(i,k) = qc3d(i,k,j)
+            qi3d_hv(i,k) = qi3d(i,k,j)
+            p3d_hv(i,k) = p3d(i,k,j)
+            pi3d_hv(i,k) = pi3d(i,k,j)
+            dz8w_hv(i,k) = dz8w(i,k,j)
+            rthraten_hv(i,k) = rthraten(i,k,j)
+         end do
+      end do
 
-      !FIND THETAV-BASED PBLH (BEST FOR DAYTIME).
-      zi=0.
-      k = kthv+1
-      IF((landsea-1.5).GE.0)THEN
-      ! WATER
-          delt_thv = 0.75
-      ELSE
-      ! LAND
-          delt_thv = 1.5
-      ENDIF
-
-      zi=0.
-      k = kthv+1
-      DO WHILE (zi .EQ. 0.)
-         IF (thetav1D(k) .GE. (minthv + delt_thv))THEN
-            zi = zw1D(k) - dz1D(k-1)* &
-              & MIN((thetav1D(k)-(minthv + delt_thv))/MAX(thetav1D(k)-thetav1D(k-1),1E-6),1.0)
-        ENDIF
-        k = k+1
-         IF (k .EQ. kte-1) zi = zw1D(kts+1) !EXIT SAFEGUARD
-      ENDDO
+      if(present(a_u_bep) .and. present(a_v_bep) .and. present(a_t_bep) .and.  &
+         present(a_q_bep) .and. present(a_e_bep) .and. present(b_u_bep) .and.  &
+         present(b_v_bep) .and. present(b_t_bep) .and. present(b_q_bep) .and.  &
+         present(b_e_bep) .and. present(dlg_bep) .and. present(dl_u_bep) .and. &
+         present(sf_bep)  .and. present(vl_bep)  .and. present(frc_urb2d)) then
```

Comment:
@ldfowler58 Because this group of variables are together, either they are all available, or not, would just testing one of them, or even just checking on flag_bep be sufficient?

Reply:
I would prefer to keep it as is to highlight that those arrays are actually optional. To simply flag_bep may not be sufficient because we may want to have flag_bep be false but the arrays to be present to ensure that results remain the same.

Reply:
@weiwangncar I agree that performance could be improved if we checked for the presence of just a single optional argument. For now, perhaps we could just let the code be as is, and work out an optimization for a future PR.

Reply:
I like Michael's suggestion for now. Thanks.